xbean-finder/src/test/java/org/apache/xbean/finder/archive/JarArchiveTest.java

```diff
@@ -113,35 +114,75 @@ public void testIterator() throws Exception {
         assertEquals(classes.length, actual.size());
     }
 
-    private List<String> list(final JarArchive archive) {
-        final List<String> actual = new ArrayList<>();
-        for (final Archive.Entry entry : archive) {
-            actual.add(entry.getName());
+    @Test
+    public void testXBEAN337() throws Exception {
+
+
+        // Virtual path
+
+        String path = "/this!/!file!/does!/!not/exist.jar";
+        URL[] urls = {new URL("jar:file:" + path + "!/some!/!inner!/.jar!/file.jar")};
+
+        try(JarArchive jar = new JarArchive(new URLClassLoader(urls), urls[0])){
+
+        }catch(Exception ex){
+            Assert.assertTrue(String.format(
+                    "Muzz never fail on '/this', but try full path with exclamations('%s') instead",
+                    path),
+                    ex.getCause().getMessage().contains("exist.jar"));
+        }
+
+
+        // Real file
+
+        File tmpDir = testTmpDir.newFolder("!" + JarArchiveTest.class.getSimpleName() + "!-temp!");
+
+        File exclamated = Files.copy(JarArchiveTest.classpath.toPath(),
+                tmpDir.toPath().resolve(
+                        JarArchiveTest.classpath.getName()))
+                .toFile();
+
+        urls[0] = new URL("jar:" + exclamated.toURI().toURL() + "!/");
+
+        try(JarArchive jar = new JarArchive(new URLClassLoader(urls), urls[0])){
+
+            Assert.assertEquals(String.format("Muzz successfully open '%s'", exclamated.getAbsolutePath()),
+                    this.archive.iterator().hasNext(),
+                    jar.iterator().hasNext());
+        }
+
+
+        // Unsupported protocols stack
+
+        urls[0] = new URL("http:ftp:jar:" + exclamated.toURI().toURL() + "!/");
+
+        try(JarArchive jar = new JarArchive(new URLClassLoader(urls), urls[0])){
+            Assert.fail(String.format("Muzz eat only local file URLs:"
+                    + " 'file:/...' or 'jar:file:/...!/' but not '%s'",
+                    urls[0]));
+        }catch(UnsupportedOperationException ex){
```

Comment:
Sorry, a late question, why does it catch `UnsupportedOperationException` instead of `IllegalArgumentException` here?

Reply:
it's my mistake, definitely IllegalArgumentException should be instead of UnsupportedOperationException.
the test fails with UnsupportedOperationException, i wonder how i could miss it.

guys, my project configuration is now broken. can i kindly ask you to include this little fix into your commits?

Reply:
```
Apache Maven 3.8.7 (b89d5959fcde851dcb1c8946a785a163f14e1e29)
Maven home: /home/rmannibucau/.sdkman/candidates/mvnd/0.9.0
Java version: 1.8.0_312, vendor: Azul Systems, Inc., runtime: /home/rmannibucau/.sdkman/candidates/java/8.0.312-zulu/jre
Default locale: fr_FR, platform encoding: UTF-8
OS name: "linux", version: "5.19.0-35-generic", arch: "amd64", family: "unix"
```

master, no diff and build is green, we should likely refine why to ensure we dont hide an issue in tests IMHO
will try to look in a few hours why it passes for me

Reply:
i just have no idea how i could commit a failing test 3 months ago... and how @rmannibucau could accept it...

but now in my(obsolete & broken) context the test fails - there is 'throw IllegalArgument..' in constructor AND 'catch UnsupportedOp...' in test

Reply:
It doesn't matter. Let's fix it together :)